### PR TITLE
Release w20

### PR DESCRIPTION
**Pull request by technight**
RELEASE WEEK 20
=============

### CODE REVIEW

File: report/4.Solution/GeneratedCppCode.tex [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+\textbf{Lists}
+
+For lists, \texttt{C++} includes a list type in its standard library called vector. 
+This vector type is, like the string type, not available in the libraries provided by the Arduino.
+Furthermore, since the vector implementation is mutable and the lists in HCL are immutable, a simple implementation of a constant list is implemented instead.
+
+One thing to be aware of when using lists, is that the size of the list cannot always be determined at compile time.
+This means that lists will usually do heap allocation. 
+With heap allocation comes the need to do proper memory handling, eg.
+clean up allocated memory in order to avoid memory leaks.
```
Comment (Abildgaard96) [inline report/4.Solution/GeneratedCppCode.tex]: clean -> Clean

File: report/4.Solution/DesignOfCompiler.tex [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+\end{figure}
+
+
+The compiler is composed of a lexer, a parser, and a code generator, as seen in figure \ref{fig:compilerProcess}.
+The lexer generates a token stream from a text stream, which is fed into the parser, which then builds an AST based on the tokens.
+The abstract syntax tree is then processed by the code generator, which outputs code in our output language, \texttt{C++}.
+
+Some compilers are built with an isolated type checker.
+However, as the HCL language only allows predefined types and has a relatively simple structure, the type checking happens as part of the parser.
+%Tobias' forslag til en mere detaljeret forklaring af sætningen på linje 19:
```
Comment (Lambcreek) [inline report/4.Solution/DesignOfCompiler.tex]: Er dette bare obsolete?

File: report/5.Conclusion/Conclusion.tex [hunk 3]
```diff
@@ -1,3 +1,78 @@
...
+\end{lstlisting}
+
+The two function calls on lines 3 and 5 both evaluate to the same value and with no ambiguity in the syntax.
+
+This finishes the conclusion of the problem-statement.
+
+\subsection{Conclusion on the solution} %Shit rhymes
+HCL is a fully fledged, Turing complete, language.
+Every part of the syntax has an equivalent semantic context. 
+The compiler is capable of compiling most syntactical constructs that are valid in HCL to C++, barring any bugs and faults in the compiler as of the writing of this report.
```
Comment (Abildgaard96) [inline report/5.Conclusion/Conclusion.tex]: think we agreed that all C++ should be \texttt{C++}

File: report/5.Conclusion/Discussion.tex [hunk 4]
```diff
@@ -1,2 +1,66 @@
 % !TeX root = ../main.tex
+\chapter{Discussion}
 \section{Discussion}
+This chapter reflects upon the choices made in regards to the developed solution and report, as well as how the group delegated work to the individual group members and assured the quality of said work.
+Furthermore, the major choices, not already touched upon, are discussed in this chapter.
+
+The goal of HCL is to develop an "English-like" language.
+The choice of an English-like language, instead of a Danish-like language, is to easier introduce programming concepts to the target-group.
+By being an English-like language, HCL expressions are designed to naturally resemble english written sentences.
```
Comment (Abildgaard96) [inline report/5.Conclusion/Discussion.tex]: english - English

File: report/5.Conclusion/Discussion.tex [hunk 5]
```diff
@@ -1,2 +1,66 @@
...
+\begin{lstlisting}[language=HCL,label=lis:discussionintuitive,firstnumber=1,caption=Example of the intuitive nature of HCL]
+func doSomething = () : none { Some Code that does stuff }
+func thisIsTrue = () : bool { Some code that returns a bool }
+
+:doSomething while :thisIsTrue
+\end{lstlisting}
+
+As discussed in the problem analysis, the majority of the target group has no understanding of operators.
+Users of HCL do not need to know the basics of logic operations as presented in discrete mathematics.
+This is because the implicit understanding of logical constructs, is natural in HCL, as the built-in operators have english names, instead of just being a character.
```
Comment (Abildgaard96) [inline report/5.Conclusion/Discussion.tex]: english -> English

File: report/5.Conclusion/Conclusion.tex [hunk 6]
```diff
@@ -1,3 +1,78 @@
...
+HCL has English-like syntax, that also implements high-order functionality.
+%However, as of the conclusion of this report, no user-tests have been conducted (This sentence is really fucking out of place. I have removed it.)
+
+The similarities with the English language is well exemplified in snippet \ref{lis:ConclusionNicolajEksempel}.
+%Decide and rewrite.
+\begin{lstlisting}[language=hcl,label=lis:ConclusionNicolajEksempel,caption=English-like syntax example]
+1 to 5 map { value squared } print
+\end{lstlisting}
+In snippet \ref{lis:ConclusionNicolajEksempel}, the values from 1 to 5 are mapped with the value squared and then printed.
+The code itself is pretty self-explanatory, and matches the english explanation a lot.
```
Comment (Lambcreek) [inline report/5.Conclusion/Conclusion.tex]: english -> English

File: report/5.Conclusion/Conclusion.tex [hunk 7]
```diff
@@ -1,3 +1,78 @@
...
+Having the first parameter on the left-hand side of the function call, allows for writing code, that reads like english.
```
Comment (Lambcreek) [inline report/5.Conclusion/Conclusion.tex]: english -> English

File: report/5.Conclusion/Discussion.tex (hunk 4 above)
Comment (Lambcreek) [inline report/5.Conclusion/Discussion.tex]: english -> English

File: report/5.Conclusion/Discussion.tex (hunk 5 above)
Comment (Lambcreek) [inline report/5.Conclusion/Discussion.tex]: english -> English

File: report/2.Analysis/ProblemStatement.tex [hunk 8]
```diff
@@ -5,12 +5,12 @@
 \section{Problem statement}
 Based on the analysis the following problem statement is defined:
 \begin{center}
-	Students have trouble understanding the basic concepts of programming, because the languages they are being taught are complicated, the reason being that they are not designed with beginners in mind
-	The group will try to solve this, by designing a simpler language, targeted at beginners.
+	\textit{Students have trouble understanding the basic concepts of programming, because the languages they are being taught are complicated, the reason being that they are not designed with novices in mind.}
+	\textit{The group will try to solve this, by designing a simpler language, targeted at novices.}
 
 \begin{itemize}
 	\item Can english-like syntax be implemented in a programming language?
```
Comment (Lambcreek) [inline report/2.Analysis/ProblemStatement.tex]: english -> English

File: report/4.Solution/DesignOfCompiler.tex (hunk 2 above)
Comment (Lambcreek) [reply report/4.Solution/DesignOfCompiler.tex]: If so, remove

Comment (nillas12) [inline report/4.Solution/DesignOfCompiler.tex]: ?????? these comments?

File: report/4.Solution/GeneratorClasses.tex [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	listOf(
+		HelperHeaders.constList,
+		HelperHeaders.ftoa,
+		FilePair("builtin.h", CodeGenerator().generate(ast.builtins())),
+		FilePair("types.h", TypeGenerator().generate(ast)),
+		FilePair("main.cpp", MainGenerator().generate(ast.addReturnCode().notBuiltins()))
+)
+\end{lstlisting}
+
+The function first creates header files for the helper classes \textit{constList}(3) and \textit{ftoa}(4), which allow the use of the list-type in \texttt{C++}, and for converting doubles to text.
```
Comment (Lambcreek) [inline report/4.Solution/GeneratorClasses.tex]: Klassen hedder "ConstList" (med stort)
ftoa er ikke en klasse - det er en funktion

File: report/4.Solution/TreeTraversal.tex [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+The code generation phase of the compiler traverses the AST and generates code according to the nodes in the AST.
+Usually, traversal of the tree is done in a depth first manner, and code is generated based on visited nodes.
+Traditionally, the AST is traversed by utilizing the visitor pattern. 
+However, more modern languages, especially functional oriented languages, have introduced the concept of pattern matching.
+Pattern matching allows for a short and concise way of traversing the tree.
+
+When possible, the pattern matching method is often preferred over the traditional visitor pattern. 
+
+Consider an AST for the following code:
+$$x = 5 + 7$$
```
Comment (Lambcreek) [inline report/4.Solution/TreeTraversal.tex]: Dobbelt dollartegn??

**Pull request merged**